:sparkles: Partitionset reconciliation

fgiloux opened the pull request:
## Summary

This PR implements a reconciliation mechanism so that shard partitions can be automatically created based on specified dimensions and selector.

[Design document](https://hackmd.io/EuPmah62QHSEs-po04Kn1Q?view)

## Related issue(s)

Follows #2469
Fixes #2334

fgiloux commented:
/retest

p0lyn0mial commented on pkg/apis/topology/v1alpha1/types_partitionset.go:
```diff
@@ -58,16 +60,47 @@ type PartitionSetSpec struct {
+	// the MatchLabels generated for the Partition according to the dimensions in the PartitionSet.
+	// This does not limit the filtering as any MatchLabel can also be expressed as
+	// a MatchExpression.
+	MatchExpressions []metav1.LabelSelectorRequirement `json:"matchExpressions,omitempty"`
```
Would it make sense to make `PartitionSetSpec.Dimensions` and/or `PartitionSetSpec.matchExpressions/PartitionSetSpec.selector` immutable ?

fgiloux replied:
As I see it this is not the likely flow.  This does not take in considerations that Partitions need to be colocated with the APIExportEndpointSlices that consume them. Here is how I see the flow:
- A service provider update a PartitionSet
- The set of Partitions bound to the PartitionSet is updated accordingly in the PartitionSet workspace
- Each of this Partition is copied into the workspace where the intended APIExportEndpointSlice is located. Each APIExportEndpointSlice is intended for an operator/controller manager instance. It makes URLs for a set of shards identified by the Partition available to it.
- The service provider can choose at any convenient time to patch the APIExportEndpointSlices with the new Partitions
There is no modification of the service and no disruption until the APIExportEndpointSlices are patched.

The steps above could be automated but it is another discussion.

p0lyn0mial replied:
I'm not sure if I read the code correctly but it seems that we are deleting old partitions potentially being used by APIExportEndpointSlice which seems to be disruptive.

xref: https://github.com/kcp-dev/kcp/blob/c991948b1e6f7b14152b0a2714930b1b1c42b4b5/pkg/reconciler/topology/partitionset_reconcile.go#L123

p0lyn0mial reviewed (commented):
I did a first pass, for now only over the API. Let me know what you think.

p0lyn0mial commented on pkg/apis/topology/v1alpha1/types_partitionset.go:
```diff
@@ -58,16 +60,47 @@ type PartitionSetSpec struct {
+	// the MatchLabels generated for the Partition according to the dimensions in the PartitionSet.
+	// This does not limit the filtering as any MatchLabel can also be expressed as
+	// a MatchExpression.
+	MatchExpressions []metav1.LabelSelectorRequirement `json:"matchExpressions,omitempty"`
```
Also could we rename it to `shardSelector/shardMatchExpression` ?

fgiloux replied:
as per the comment " MatchExpressions are used here instead of Selector to avoid possible collisions with the MatchLabels generated for the Partition".
PartitionSets are used to generate Partitions. Partitions have a LabelSelector. LabelSelectors have two parts:
-  MatchLabels
- MatchExpressions

MatchLabels of Partitions are generated from the Dimension field of PartitionSet. MatchExpressions of Partitions are copied from MatchExpressions of PartitionSet. Keeping  this separation provides a clearer UX. It would be more difficult for users to understand the generated Partitions from a PartitionSet if MatchLabels could be a mix of  the Dimension generation and MatchLabels inherited from the PartitionSet. As per the comment there is no trade-off in terms of capabilities: any MatchLabels can be expressed by MatchExpressions.

p0lyn0mial replied:
> as per the comment " MatchExpressions are used here instead of Selector to avoid possible collisions with the MatchLabels generated for the Partition".

if we have a collision, does it mean we can achieve the same by providing a selector on a `Partition` resource :) ?

fgiloux replied:
If I understand your point correctly PartitionSet is just a convenient way to create Partitions. Service Providers are free to create Partitions the way it best suits them: using PartitionSet or directly creating the resource. If they do it manually they can set Selectors: MatchLabels and MatchExpressions in Partitions in a similar way to what is generated by the controller.

fgiloux commented on pkg/reconciler/topology/partitioning_test.go:
```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2022 The KCP Authors.
```
The PR and commit were created in 2022

p0lyn0mial commented on pkg/reconciler/topology/resources.go:
```diff
@@ -0,0 +1,39 @@
+func generatePartition(name string, matchExpressions []metav1.LabelSelectorRequirement, matchLabels map[string]string) *topologyv1alpha1.Partition {
+	return &topologyv1alpha1.Partition{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: name + "-",
```
It appears to me that providers will have to manually introspect generated `Partitions` to make use of them because names appear random and there is no explanation of the generated selector, am I right?

Wondering if we could/should improve this?

fgiloux replied:
Names are generated as Partitions are. We could create deterministic names by hashing the Partition spec but I am not sure it would be reasonable. An annotation may be better suited to store the hash. This would be stable as the PartitionSet reconciliation only does Partition create/delete, no update.
That said I am not sure it would be a significant improvement:
- Shard topology should be stable, e.g. you may add new shards in a region time to time but adding a new region is a rare event that would need action from Service Providers for them to cover it. It would also mean just a few additional Partitions to be taken care of. Deletion is not much different.
- PartitionSet change is driven by Service Provider and it means that they want to change the distribution of their service deployments. In that case they can check the deltas between the deployed Partitions and the newly generated one, even if a random name is used for Partitions. Some tooling on the client side may help with that.
There may be bits in the workflow that can be improved and I am open to proposals. My current thinking is that this could get better addressed when we develop some client side tooling that automates service providers deployments and topology changes, possibly taking a PartitionSet and an APIExport as parameter. I am a bit wary with having this automation on the server side because:
- that could drive automated changes that introduce big disruption and cost increases by service providers, so that it would need to be gated, hence may not bring much gain compared to a client-side approach.
- I am not sure that cross-shard communication will always be possible.

fgiloux commented:
/retest

sttts commented on pkg/reconciler/topology/resources.go:
```diff
@@ -22,12 +22,20 @@ import (
 	topologyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/topology/v1alpha1"
 )
 
+const dimensionsLabelKeyBase = "partition.kcp.io/"
```
`partition.topology.kcp.io/`

fgiloux commented:
@sttts @p0lyn0mial I have implemented the changes we discussed:
- using the dimension values for Partition names (still using name generator taking care of overflows)
- removed the labels with the dimensions (as the information is now made visible by the name)
- using a selector in PartitionSet rather than matchExpressions only (some collision check was needed)

fgiloux commented:
/retest

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+func (c *controller) deletePartitions(ctx context.Context, partitionSet *topologyv1alpha1.PartitionSet) error {
+	partitionSet.Status.Count = 0
```
oh, it tracks the number of associated `partitions`.
should we then decrease this number by the number of successful calls to `deletePartition` ?

fgiloux replied:
`deletePartitions` (mind the `s` at the end) is only call in case of an issue with the selector (lines 55 and 83) the count should then be set back to 0. In case of regular deletion because of a change in shards `deletePartition`without `s` is called (lines 147 and 182). In case of no error with the selector the number of endpoints is already captured in the size of `matchLabelsMap`, no need for additional counting.
I will move `partitionSet.Status.Count = 0`one level up as it is not clear from the name of the function that it does that as well.

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+	}
+
+	var matchLabelsMap map[string]map[string]string
+	if partitionSet.Spec.ShardSelector != nil {
```
it looks like we could get rid of this `if` statement since the `partition` function can already handle nil `shardSelectorLabels`

fgiloux replied:
I cannot reference `MatchLabels` on a nil pointer

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+
+// partition populates shard label selectors according to dimensions.
+// It only keeps selectors that have at least one match.
+func partition(shards []*corev1alpha1.Shard, dimensions []string, shardSelectorLabels map[string]string) (matchLabelsMap map[string]map[string]string) {
```
would it make sense to rename to `partitionByDimensionsAndLabels`?

fgiloux replied:
I don't see the need for now. There is currently a single way of partitioning. `Labels`could also be ambiguous in this context. `Dimensions`are a way to slice and dice by `Labels`. `Labels`are also used for filtering, defining the boundaries of the piece to slice and dice. What I mean is that `Dimensions`and `Labels`are not of the same nature for the partitioning.

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+		for _, label := range labels {
+			labelValue, ok := shard.Labels[label]
+			if !ok {
+				break
```
what does it mean?
what if we didn't find let's say the second label out of four? (`len(key) > 0` will apply)

fgiloux replied:
true. Adding a boolean to check the match

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
```
again, since this is a network call, we could be more greedy and try all `oldPartitions`, wdyt?

fgiloux replied:
I am not sure how you would do that here. There is no single selector for the Partitions to delete.

fgiloux commented on pkg/server/server.go:
```diff
@@ -453,6 +453,12 @@ func (s *Server) Run(ctx context.Context) error {
 		}
 	}
 
+	if s.Options.Controllers.EnableAll || enabled.Has("partition") {
```
I would be ok with changing to partitionset. I used partition as we don't have any other controller specific to Partitions. Partitions alone don't get reconciled. It is only in the context of an APIExportEndpointSlice that the controller takes the referenced Partition in consideration for populating  the endpoints. The same mechanism could also be leveraged in other places in the future.

sttts replied:
we should honestly remove all these ifs. We don't use them anywhere.

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+			if !ok {
+				break
+			}
+			key = key + "+" + label + "=" + labelValue
```
i'm not sure if i understand why the `key` has to collect all labels with their values, couldn't we calculate some `hash` from the labels and values ?

fgiloux replied:
This is a kind of hash. We could use a hash function for it. I am not sure that the processing time would be shorter and it was nice to have something readable by debugging.

p0lyn0mial replied:
if we won't reduce the size of the key it might grow indefinitely, no?

fgiloux replied:
The keys are only used as a transient and quick way of comparing the desired state of Partitions with the existing ones. They are not stored.

fgiloux commented on pkg/reconciler/topology/partitionset_controller.go:
```diff
@@ -0,0 +1,346 @@
+		return
+	}
+
+	logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), obj.(*topologyv1alpha1.Partition))
```
done

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+					conditionsv1alpha1.ConditionSeverityError,
+					"partition could not get created",
+				)
+				return err
```
since this is a network call, we could be more greedy and try to create all `partitions`

fgiloux replied:
I am not sure how you would do that

fgiloux commented on pkg/reconciler/topology/partitionset_indexes.go:
```diff
@@ -0,0 +1,50 @@
+	for _, ownerRef := range partition.OwnerReferences {
+		if ownerRef.Kind == "PartitionSet" {
+			path := logicalcluster.From(partition).Path()
+			if !ok {
```
done

p0lyn0mial reviewed (commented):
I did a first pass, let's see if we can make move some parts into an admission plugin. I think that would simplify the controller. Let me know what you think, thanks!

fgiloux commented on pkg/reconciler/topology/partitionset_controller.go:
```diff
@@ -0,0 +1,346 @@
+	if !ok {
+		return false
+	}
+	if !reflect.DeepEqual(oldShard.Labels, newShard.Labels) {
```
return !reflect.DeepEqual(oldShard.Labels, newShard.Labels)
done

fgiloux commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+)
+
+func (c *controller) reconcile(ctx context.Context, partitionSet *topologyv1alpha1.PartitionSet) error {
+	if ok, err := isSelectorWithoutConflict(partitionSet.Spec.ShardSelector, partitionSet.Spec.Dimensions); !ok {
```
We could. I found the logic to complicate for CEL and I tend to avoid admission plugins for non-security related checks but yes I could do that if you think that it improves the UX.

p0lyn0mial replied:
we could move it either to CEL or an admission plugin, it seems like we could also move https://github.com/kcp-dev/kcp/pull/2513#discussion_r1091111068. 
I think it would simply this controller and would avoid deleting `partitions`

@sttts wdyt?

fgiloux commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error listing Shards",
```
The error is returned by the reconcile function, hence logged by the controller

fgiloux commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+		labelSelector = labels.Everything()
+	} else {
+		var err error
+		labelSelector, err = metav1.LabelSelectorAsSelector(partitionSet.Spec.ShardSelector)
```
see above. We can discuss it on slack

fgiloux commented on pkg/reconciler/topology/partitionset_controller.go:
```diff
@@ -0,0 +1,346 @@
+		getPartitionSet: func(clusterName logicalcluster.Name, name string) (*topologyv1alpha1.PartitionSet, error) {
+			return partitionSetClusterInformer.Lister().Cluster(clusterName).Get(name)
+		},
+		getPartitionsByPartitionSet: func(key string) ([]*topologyv1alpha1.Partition, error) {
```
yes I can change the signature of the method but still we will need to calculate the key in the method to use the indexer, won't we?

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,278 @@
+	}
+	for _, partition := range existingPartitions {
+		if err := c.deletePartition(ctx, logicalcluster.From(partition).Path(), partition.Name); err != nil && !apierrors.IsNotFound(err) {
+			return err
```
just loop over `existingPartitions`, try to `deletePartition`, collect any errors and return at the end :) ?

fgiloux replied:
ok so it makes no difference in terms of network calls, does it?

p0lyn0mial replied:
It does, we try all `existingPartitions` even if we failed to delete a `partition`.  Hopefully on the next iteration we will have less work do to. Does it make sense?

sttts commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,281 @@
+			topologyv1alpha1.PartitionSetValid,
+			topologyv1alpha1.PartitionSetInvalidSelectorReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
```
this needs context

sttts commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,281 @@
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"",
```
this needs context

fgiloux commented:
@sttts @p0lyn0mial as discussed:
- I removed the check when matchLabels and dimensions overlap.
- I have added openapi/CEL validation for label selector.

fgiloux commented:
/retest

fgiloux commented:
/retest

fgiloux commented:
/retest

fgiloux commented:
/retest

sttts commented:
/retest

fgiloux commented:
/retest

ncdc commented:
#2665 
/retest

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,242 @@
+}
+
+// partition populates shard label selectors according to dimensions.
+// It only keeps selectors that have at least one match.
```
with https://github.com/kcp-dev/kcp/pull/2513/files#r1091643038 this comment is not true, correct?

fgiloux replied:
I will make the comment clearer, what I meant: "It only keeps selectors that have at least one Shard matching them so that Partition not referring to any Shards should not get created"

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,242 @@
+	labels := make([]string, len(dimensions), len(dimensions)+len(shardSelectorLabels))
+	copy(labels, dimensions)
+	for label := range shardSelectorLabels {
+		labels = append(labels, label)
```
shouldn't we also take values into account ?

fgiloux replied:
If I am following properly: The Shards passed to the `partition` function is a subset of all the Shards that comply with the Selector configured on PartitionSet.

p0lyn0mial replied:
okay, so the values will be taken into okay just on different level, thx.

p0lyn0mial commented on pkg/reconciler/topology/resources.go:
```diff
@@ -0,0 +1,49 @@
+	copy(labels, dimensions)
+	sort.Strings(labels)
+	for _, label := range labels {
+		pname = pname + "-" + matchLabels[label]
```
are there any name length limit ?
what if we hit one ?

fgiloux replied:
there are name limits. GenerateName takes care of it

p0lyn0mial replied:
ok, thx.

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,209 @@
+)
+
+func (c *controller) reconcile(ctx context.Context, partitionSet *topologyv1alpha1.PartitionSet) error {
+	var labelSelector labels.Selector
```
nit: we could get rid of the `else` clause by doing
```
labelSelector := labels.Everything()
	if partitionSet.Spec.ShardSelector != nil {...}
```

fgiloux replied:
changed

p0lyn0mial commented on pkg/reconciler/topology/partitionset_reconcile.go:
```diff
@@ -0,0 +1,209 @@
+		if oldPartition.Spec.Selector != nil {
+			oldMatchExpressions = oldPartition.Spec.Selector.MatchExpressions
+		}
+		if !equality.Semantic.DeepEqual(oldMatchExpressions, newMatchExpressions) {
```
would it make sense to change this code so that there is only one path that deletes a partition ? i.e.
```
if  !matchExpressions || !partitionKey {
 delete...
}
```

fgiloux replied:
`deletePartition` and `conditions.MarkFalse`are two lines. If I understand your point correctly, in term of readability I prefer to have the action next to the evaluation rather than differing it.

p0lyn0mial commented:
@fgiloux I'd also put the new controller into a new directory i.e. `topology/partitionset/` I think atm it is under `topology/`

p0lyn0mial commented on config/crds/topology.kcp.io_partitionsets.yaml:
```diff
@@ -66,11 +65,26 @@ spec:
+                          maxLength: 317
                           type: string
+                          x-kubernetes-validations:
+                          - message: Label key prefixes are optional and limited to
```
I have no way of verifying this validation, I think we will validate it in an e2e test.

fgiloux replied:
yes I will add checks for the admission to the e2e PR

fgiloux commented:
> @fgiloux I'd also put the new controller into a new directory i.e. `topology/partitionset/` I think atm it is under `topology/`

I can do that. It is currently the only controller in topology

fgiloux commented:
/retest

sttts commented on config/root-phase0/apiresourceschema-partitionsets.topology.kcp.io.yaml:
```diff
@@ -47,8 +47,8 @@ spec:
-            selector:
-              description: selector (optional) is a label selector that filters shard
+            shardSelector:
+              description: ShardSelector (optional) specifies filtering for shard
```
lower-case

sttts commented on pkg/apis/topology/v1alpha1/types_partitionset.go:
```diff
@@ -58,16 +60,43 @@ type PartitionSetSpec struct {
 
-	// selector (optional) is a label selector that filters shard targets.
-	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+	// ShardSelector (optional) specifies filtering for shard targets.
```
lower-case

sttts commented on pkg/reconciler/topology/partitionset/partitionset_controller.go:
```diff
@@ -0,0 +1,346 @@
+)
+
+const (
+	ControllerName = "kcp-partitionset"
```
kcp-topology-partitionset

sttts commented on pkg/reconciler/topology/partitionset/partitionset_controller.go:
```diff
@@ -0,0 +1,346 @@
+			if err != nil {
+				return nil, err
+			}
+			list, err := partitionClusterInformer.Informer().GetIndexer().ByIndex(indexPartitionsByPartitionSet, key)
```
I think we have a helper somewhere for this code, using generics.

sttts commented on pkg/reconciler/topology/partitionset/partitionset_controller.go:
```diff
@@ -0,0 +1,346 @@
+				c.enqueueAllPartitionSets(obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if filterShardEvent(oldObj, newObj) {
```
explain in a comment what the filtering is and why it makes sense

sttts commented on pkg/reconciler/topology/partitionset/partitionset_controller.go:
```diff
@@ -0,0 +1,346 @@
+
+	indexers.AddIfNotPresentOrDie(partitionClusterInformer.Informer().GetIndexer(), cache.Indexers{
+		indexPartitionsByPartitionSet: indexPartitionsByPartitionSetFunc,
+	})
```
nit: indexes first, event handler after that.

sttts commented on pkg/reconciler/topology/partitionset/partitionset_controller.go:
```diff
@@ -0,0 +1,346 @@
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	// kcpClusterClient is cluster aware and used to communicate with kcp API servers
```
no need for the comment. This is convention.

sttts commented on pkg/reconciler/topology/partitionset/partitionset_controller.go:
```diff
@@ -0,0 +1,346 @@
+		listShards: func(selector labels.Selector) ([]*corev1alpha1.Shard, error) {
+			return globalShardClusterInformer.Lister().List(selector)
+		},
+		listPartitionSets: func() ([]*topologyv1alpha1.PartitionSet, error) {
```
this looks suspicious. You should never need to list all partition sets.

sttts replied:
Nvmd. I understand why you need it.

sttts commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+
+func (c *controller) reconcile(ctx context.Context, partitionSet *topologyv1alpha1.PartitionSet) error {
+	labelSelector := labels.Everything()
+	if partitionSet.Spec.ShardSelector == nil {
```
!= nil ?

sttts commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+	}
+	partitionSet.Status.Count = uint16(len(matchLabelsMap))
+	existingMatches := map[string]struct{}{}
+	var placeHolder struct{}
```
just use boolean type for the map

sttts commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+		matchLabelsMap = partition(shards, partitionSet.Spec.Dimensions, nil)
+	}
+	partitionSet.Status.Count = uint16(len(matchLabelsMap))
+	existingMatches := map[string]struct{}{}
```
this can be made type-safe. Compare https://goplay.space/#IxSJ1_iWaY9.

sttts replied:
I meant the key, i.e. that the key is a string but could be a struct. No need to think about representation.

sttts commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+		if _, ok := matchLabelsMap[partitionKey]; ok {
+			existingMatches[partitionKey] = placeHolder
+		} else {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil {
```
&& !IsNotFound

fgiloux replied:
added

sttts commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+			}
+		}
+	}
+	// Create partitions when no existing partition for the set has the same selector.
```
nit: empty line above

fgiloux replied:
added

sttts commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+		if _, ok := matchLabelsMap[partitionKey]; ok {
+			existingMatches[partitionKey] = placeHolder
+		} else {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil {
```
We usually put `logger.V(2)` before every client call

fgiloux replied:
added

sttts commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+			partition.OwnerReferences = []metav1.OwnerReference{
+				*metav1.NewControllerRef(partitionSet, topologyv1alpha1.SchemeGroupVersion.WithKind("PartitionSet")),
+			}
+			_, err = c.createPartition(ctx, logicalcluster.From(partitionSet).Path(), partition)
```
logger output

fgiloux replied:
added

sttts commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+		labels = append(labels, label)
+	}
+	// Sorting for consistent comparison.
+	sort.Strings(labels)
```
```
=== empty line ===
// Sorting for consistent comparison.
sort.Strings(labels)
=== empty line ===
```
or 
```
sort.Strings(labels) // Sorting for consistent comparison.
```

fgiloux replied:
changed

sttts commented on pkg/reconciler/topology/partitionset/resources.go:
```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2022 The KCP Authors.
```
2023

sttts replied:
and everywhere else

fgiloux replied:
I created the PR in 2022. Anyways I can change that if it matters

sttts commented on config/crds/topology.kcp.io_partitionsets.yaml-patch:
```diff
@@ -0,0 +1,58 @@
+- op: add
+  path: /spec/versions/name=v1alpha1/schema/openAPIV3Schema/properties/spec/properties/shardSelector/properties/matchExpressions/items/properties/values/items/pattern
+  value:
+    ^[A-Za-z0-9]([-A-Za-z0-9_.]{0,61}[A-Za-z0-9])?$
```
@vincepri label selector validation should come from controller-tools, shouldn't it?

fgiloux commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+	if partitionSet.Spec.ShardSelector != nil {
+		newMatchExpressions = partitionSet.Spec.ShardSelector.MatchExpressions
+	}
+	for _, oldPartition := range oldPartitions {
```
added

fgiloux commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -0,0 +1,208 @@
+
+		// Sorting the keys of the old partition for consistent comparison
+		sortedOldKeys := make([]string, len(oldMatchLabels))
+
```
removed

fgiloux commented:
/retest

p0lyn0mial commented:
/lgtm

thanks!

sttts commented:
/approve
